MESOS: Decouple executor and kubelet

sttts opened the pull request:
This patch reduces the dependencies of the executor from the kubelet. This makes it possible to launch the kubelet after the executor.

This PR is the pre-requisite for
- passing of slave resources as node capacity to the kubelet: https://github.com/kubernetes/kubernetes/pull/14260
- launching an unmodified, standalone kubelet some day.

All non-mesos changes are moved into https://github.com/kubernetes/kubernetes/pull/14415

sttts committed:
Remove redundant kubelet dependency of executor

sttts committed:
Un-knot static pods creation in executor

sttts committed:
Consolidate executor driver initialization code

sttts committed:
Decouple executor initialization from kubelet

This patch reduces the dependencies of the executor from the kubelet. This
makes it possible launch the kubelet after the executor.

This considerably reduces the complexity of the startup code.

Moreover, this work is a requirement to use a standalone kubelet some day.

bgrant0607 commented:
cc @cjcullen @a-robinson re. cloudprovider
cc @dchen1107 re. kubelet
cc @davidopp re. mesos

sttts commented:
@jdef this one would be ready for a high-level review. It certainly needs some clean up here and there.

jdef commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -30,14 +30,15 @@ mesosslave1:
-  entrypoint: [ "bash", "-c", "wrapdocker mesos-slave --hostname=$(getent hosts mesosslave1 | cut -d' ' -f1 | sort -u | tail -1)" ]
   command: ~
   environment:
+  - MESOS_HOSTNAME=mesosslave1
```
is this because the new dind image supports this variable?

sttts replied:
No, it was supported before, but we did not explicitly pass it to the kubelet (only indirectly as hostname). In production environment one might have real DNS names for slaves, not IPs. That's why I added it here.

jdef commented on cluster/mesos/docker/km/build.sh:
```diff
@@ -67,7 +67,7 @@ echo "Copying files to workspace"
 #cp "${script_dir}/bin/"* "${workspace}/bin/"
 cp "${common_bin_path}/"* "${workspace}/bin/"
-cp "${kube_bin_path}/"* "${workspace}/bin/"
+cp "${kube_bin_path}/km" "${workspace}/bin/"
```
we may need to revert this if we ever decide to serve kubectl from the image

sttts replied:
This is only our docker-compose cluster for now. But of course you are right if we want to test such a kubectl serving functionality in this setup.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -125,16 +117,16 @@ type KubernetesExecutor struct {
+	podStatusFunc        func(*api.Pod) (*api.PodStatus, error)
 	staticPodsConfigPath string
 	initialRegComplete   chan struct{}
+
```
nit: fix formatting
- remove blank line
- move comment to right side of the variable
- remove the space after the asterisk

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -320,10 +409,23 @@ func (ks *KubeletExecutorServer) createAndInitKubelet(
 		MaxContainers:      kc.MaxContainerCount,
 	}
 
+	// create main pod source, emit first empty update and pipe executor updates into it
```
"emit first empty update" -- yes, that's needed but I don't see it

sttts replied:
Good spot. Moved the code but not comment.

jdef commented on contrib/mesos/pkg/scheduler/podtask/pod_task.go:
```diff
@@ -211,6 +193,24 @@ func (t *T) Reset() {
 	t.Spec = Spec{}
 }
 
+// SlaveLabels converts slave attributes into string key/value labels
```
can we pull this out of pod_task.go?

sttts replied:
Yeah. Had the same thought.

jdef commented on contrib/mesos/pkg/scheduler/podtask/pod_task.go:
```diff
@@ -211,6 +193,24 @@ func (t *T) Reset() {
+			continue
+		}
+
+		switch a.GetType() {
```
any value in trying to support SET types? they're just comma delimited lists, and k8s has some support for set-based field selectors. not sure how frequently they're applied to nodes that have label values containing sets.

sttts replied:
The actual labels in the Node spec are string only.

jdef replied:
right, but i was wondering if k8s ever made use of label values that were CSV

jdef commented on contrib/mesos/pkg/scheduler/podtask/registry.go:
```diff
@@ -186,7 +186,7 @@ func (k *inMemoryRegistry) UpdateStatus(status *mesos.TaskStatus) (*T, StateType
 	case mesos.TaskState_TASK_FINISHED:
 		k.handleTaskFinished(task, state, status)
-	case mesos.TaskState_TASK_FAILED:
+	case mesos.TaskState_TASK_FAILED, mesos.TaskState_TASK_ERROR:
```
there's a separate PR for this work, yes? was this needed to workaround a testing blocker?

sttts replied:
Yes, this PR must be split. For CI testing I merged several pending changes.

jdef commented on pkg/cloudprovider/cloud.go:
```diff
@@ -113,6 +113,8 @@ type Instances interface {
 	// On most clouds (e.g. GCE) this is the hostname, so we provide the hostname
 	CurrentNodeName(hostname string) (string, error)
+	// Labels returns cloud provider node labels of the specified instance.
+	Labels(name string) (map[string]string, error)
```
this should be split into a separate PR

sttts replied:
Yes, for sure.

jdef commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -184,6 +199,41 @@ func (c *MesosCloud) ExternalID(instance string) (string, error) {
 
+// Labels returns the cloud provider node labels of the specified instance.
+func (c *MesosCloud) Labels(name string) (map[string]string, error) {
+	// if we MesosMaster is set, we are on a service node and query
```
s/if we/if/

jdef commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -55,31 +60,37 @@ type MesosCloud struct {
+			log.V(1).Infof("failed to create master detector: %v", err)
+			return nil, err
+		}
+
```
nit: remove blank line

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -212,7 +215,90 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+	// initialize driver and initialize the executor with it
+	dconfig := bindings.DriverConfig{
+		Executor:         exec,
+		HostnameOverride: s.HostnameOverride,
```
hostnameOverride used to contain the slave's hostname (because podtask was filling in this value based on the hostname in the offer details). this change affects the driver binding as well as the kubelet (which also consumed hostnameoverride). GetHostIP (used in PodStatusFunc) depends on this value indirectly (if i remember correctly). i'm somewhat scared of the impact here.

sttts replied:
This is exactly the part where I need your critical eyes whether the restructuring broke anything ;-) I tried to avoid any behaviour change while refactoring that. But maybe I missed some implicit dependency.

This work is also a preparation for a real split of the executor and the kubelet processes. So we have to decouple those initializations.

jdef replied:
the node name situation is kind of a mess. will need to carefully review
the current state of the kubelet code.

On Mon, Aug 24, 2015 at 1:20 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r37776231>
> :
>
> > +			status.Phase = kubelet.GetPhase(&pod.Spec, status.ContainerStatuses)
> > +			hostIP, err := s.klet.GetHostIP()
> > +			if err != nil {
> > +				log.Errorf("Cannot get host IP: %v", err)
> > +			} else {
> > +				status.HostIP = hostIP.String()
> > +			}
> > +			return status, nil
> > +		},
> > +		StaticPodsConfigPath: staticPodsConfigPath,
> > +	})
> > +
> > +	// initialize driver and initialize the executor with it
> > +	dconfig := bindings.DriverConfig{
> > +		Executor:         exec,
> > +		HostnameOverride: s.HostnameOverride,
>
> This is exactly the part where I need your critical eyes whether the
> restructuring broke anything ;-) I tried to avoid any behaviour change
> while refactoring that. But maybe I missed some implicit dependency.
>
> This work is also a preparation for a real split of the executor and the
> kubelet processes. So we have to decouple those initializations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r37776231>.
>

sttts replied:
If the standalone kubelet supports `--hostname-override` I tend to remove the hostname overriding from the cloud provider again. We don't need both ways.

jdef replied:
yes standalone kubelet supports hostname override.

On Mon, Aug 24, 2015 at 2:09 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r37782043>
> :
>
> > +			status.Phase = kubelet.GetPhase(&pod.Spec, status.ContainerStatuses)
> > +			hostIP, err := s.klet.GetHostIP()
> > +			if err != nil {
> > +				log.Errorf("Cannot get host IP: %v", err)
> > +			} else {
> > +				status.HostIP = hostIP.String()
> > +			}
> > +			return status, nil
> > +		},
> > +		StaticPodsConfigPath: staticPodsConfigPath,
> > +	})
> > +
> > +	// initialize driver and initialize the executor with it
> > +	dconfig := bindings.DriverConfig{
> > +		Executor:         exec,
> > +		HostnameOverride: s.HostnameOverride,
>
> If the standalone kubelet supports --hostname-override I tend to remove
> the hostname overriding from the cloud provider again. We don't need both
> ways.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r37782043>.
>

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -212,7 +215,90 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+	}
+	cloud, err := mcloud.New(&mcloud.Config{
+		Mesos_Node: mcloud.Node{
+			Name:   si.GetHostname(),
```
does providing the slave hostname here take the place of providing hostname override WRT the above concerns?

sttts replied:
Yes, this is kind of redundant if the standalone kubelet also supports the override-hostname flag.

sttts replied:
The kubelet internally has the Hostname and the Nodename. I think we don't need the difference in the mesos case because the Nodename concept does not exist for a slave.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -212,7 +215,90 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+	cloud, err := mcloud.New(&mcloud.Config{
+		Mesos_Node: mcloud.Node{
+			Name:   si.GetHostname(),
+			Labels: l,
```
injecting the labels here is very nice.

sttts replied:
I expect a discussion whether labels should be passed as flag or via cloud provider. We are fine with either way.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -212,7 +215,90 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+		},
+		ExitFunc: os.Exit,
+		PodStatusFunc: func(pod *api.Pod) (*api.PodStatus, error) {
+			if s.klet == nil {
```
is this even possible? if so it's data-racy

sttts replied:
Good spot. We could add a lock. But maybe we have to think about how to do this when executor and kubelet processes are split. Maybe we can also use directly the apiserver client. No need to call into the kubelet.

jdef commented on contrib/mesos/pkg/scheduler/podtask/pod_task.go:
```diff
@@ -211,6 +193,24 @@ func (t *T) Reset() {
 }
 
+// SlaveLabels converts slave attributes into string key/value labels
+func SlaveLabels(attrs []*mesos.Attribute) map[string]string {
```
would a more generic name be appropriate here? `AttributesToLabels`?

jdef commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -75,8 +75,8 @@ const (
 
-	executorCPUs = mresource.CPUShares(0.25) // initial CPU allocated for executor
-	executorMem  = mresource.MegaBytes(64.0) // initial memory allocated for executor
+	ExecutorCPUs = mresource.CPUShares(0.01) // initial CPU allocated for executor, intentionally set to the minimum allow by Mesos: 0.01
```
is this really enough for production use? with large numbers of pods running on a host and given that kubelet is also running cadvisor and constantly pinging docker containers for status .. a value of `0.01` seems quite insufficient. I suspect that the older value `0.25` would also prove insufficient at certain scales.

if this was done for testing, then I think we need to do a couple of things:
- document that fact here (and also, perhaps, in the known issues doc?)
- allow this default value to be overridden at runtime so that people doing real work w/ this framework don't suffer

if this was not done for testing then I recommend leaving value alone.

sttts replied:
This is currently the only way to pass the https://github.com/mesosphere/kubernetes-mesos/issues/437 e2e test. Can you create a GH issue about customizing this value?

jdef replied:
https://github.com/mesosphere/kubernetes-mesos/issues/456

On Mon, Aug 24, 2015 at 1:15 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r37775712>
> :
>
> > @@ -75,8 +75,8 @@ const (
> >  	defaultReconcileCooldown = 15 * time.Second
> >  	defaultFrameworkName     = "Kubernetes"
> >
> > -	executorCPUs = mresource.CPUShares(0.25) // initial CPU allocated for executor
> > -	executorMem  = mresource.MegaBytes(64.0) // initial memory allocated for executor
> > +	ExecutorCPUs = mresource.CPUShares(0.01) // initial CPU allocated for executor, intentionally set to the minimum allow by Mesos: 0.01
>
> This is currently the only way to pass the mesosphere/kubernetes-mesos#437
> <https://github.com/mesosphere/kubernetes-mesos/issues/437> e2e test. Can
> you create a GH issue about customizing this value?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r37775712>.
>

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -208,23 +202,36 @@ func (k *KubernetesExecutor) isDone() bool {
 	}
 
-	k.initialRegistration.Do(k.onInitialRegistration)
+	k.slaveInfo = slaveInfo
```
is the slaveInfo expected to change across calls to Registered and Reregistered? If not, then we're ok. If so .. then two problems:
- the mesos cloud config may become out of date w/ respect to the new info
- if Reregistered updates slaveInfo, then SlaveInfo() becomes data-racy

sttts replied:
It doesn't update the slaveInfo right now. And I am not sure there is any value in this.

jdef commented on contrib/mesos/pkg/executor/service/cadvisor.go:
```diff
@@ -0,0 +1,65 @@
+	return &MesosCadvisor{c, si}, nil
+}
+
+func (mc *MesosCadvisor) MachineInfo() (*cadvisorApi.MachineInfo, error) {
```
if slaveInfo is expected to never change then can we just do this extraction once instead of upon every invocation?

sttts replied:
Yes. But I expect that also kubelet only calls this once.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -208,23 +202,36 @@ func (k *KubernetesExecutor) isDone() bool {
 }
 
+func (k *KubernetesExecutor) SlaveInfo() *mesos.SlaveInfo {
+	return k.slaveInfo
```
this is totally in line with how the rest of the framework abuses protobufs. I'd love for it to stop. my preference here would be to declare a struct in this package:
```go
type SlaveInfo struct {
  Hostname string
  Labels map[string]string
  Resources api.ResourceList
}
```
and build one of these in Registered, and return it from SlaveInfo(). This has the nice side-affect of stopping the protobuf bleeding into the cadvisor helper code.

jdef commented on pkg/cloudprovider/providers/mesos/config.go:
```diff
@@ -35,17 +37,30 @@ const (
 //	state-cache-ttl     = 1h
+//
+// [mesos-node]
+//  labels = {"rack": "a", "gen": "2014"}
```
this comment looks out of sync w/ respect to reality

sttts replied:
good spot.

jdef commented on pkg/cloudprovider/providers/mesos/config.go:
```diff
@@ -35,17 +37,30 @@ const (
-type ConfigWrapper struct {
-	Mesos_Cloud Config
+type Config struct {
+	Mesos_Cloud Cloud
```
dislike names with both camel-case and underscores. can we use a `gcfg:..` attribute to avoid this?

jdef replied:
yes, i realize the original code had this too .. but this seems like a good opportunity to fix.

sttts replied:
I believe that I read in the gcfg docs that this is mandatory. But maybe I am wrong.

jdef commented on pkg/kubelet/kubelet.go:
```diff
@@ -772,6 +772,14 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 		if err != nil {
 			return nil, err
 		}
+
```
this should be a separate PR. i'd be tempted to group it with the Labels cloud cloud provider change, but worry that it might get nixed that way: this might be best as its own PR (and make a reference to the cloudprovider Labels PR so that people have some context)

jdef commented on test/e2e/mesos.go:
```diff
@@ -0,0 +1,96 @@
+	"k8s.io/kubernetes/pkg/fields"
+)
+
+var _ = Describe("Mesos", func() {
```
:+1:

bgrant0607 commented:
Assigning to @davidopp to find reviewers.

jdef commented:
not sure why buildbot assigned to you @bgrant0607 but this is still under internal review.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -208,23 +210,74 @@ func (k *KubernetesExecutor) isDone() bool {
+	k.nodeInfoLock.Lock()
+	defer k.nodeInfoLock.Unlock()
+
+	// clone k.nodeInfo
```
use api.Scheme.DeepCopy?

sttts replied:
Does that work with custom structs? I thoughts it uses pregenerated deep copy functions.

jdef replied:
you're right, was thinking about cloning api objects which NodeInfo is not.

On Tue, Sep 8, 2015 at 9:45 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r38926443>
> :
>
> > @@ -208,23 +210,74 @@ func (k *KubernetesExecutor) isDone() bool {
> >  	}
> >  }
> >
> > +func (k *KubernetesExecutor) NodeInfo() (ni NodeInfo) {
> > +	k.nodeInfoLock.Lock()
> > +	defer k.nodeInfoLock.Unlock()
> > +
> > +	// clone k.nodeInfo
>
> Does that work with custom structs? I thoughts it uses pregenerated deep
> copy functions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r38926443>.
>

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -208,23 +210,74 @@ func (k *KubernetesExecutor) isDone() bool {
+		// because ExecutorCPU is only set to the minumm allowed of 0.01 and
+		// it's only about shares anyway, no hard cpu limit.
+		case "mem":
+			k.nodeInfo.Mem = int64(r.GetScalar().GetValue()-float64(sservice.DefaultExecutorMem)) * 1024 * 1024
```
instead of defaultExecutorMem, can't we look at the "mem" resource of the ExecutorInfo?

sttts replied:
oh, yes. This solves the TODO above. thanks.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -288,8 +381,20 @@ func (ks *KubeletExecutorServer) createAndInitKubelet(
 	pc := kconfig.NewPodConfig(kconfig.PodConfigNotificationIncremental, kc.Recorder)
 	updates := pc.Channel(MESOS_CFG_SOURCE)
+	go func() {
+		for u := range execUpdates {
```
can we get rid of KubeletExecutor.done? add a `defer close(someDoneChan)` here since `execUpdates` closes upon termination?

sttts replied:
but KubeletExecutor.done is used everywhere in executor.go. Passing it through the whole control flow is not much nicer.

sttts replied:
I have removed executor.done, made it private as executor.terminate and instead react on the updatesChan to close to shutdown the kubelet. Feels cleaner now.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -178,7 +182,93 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+			}
+		},
+		ExitFunc: os.Exit,
+		PodStatusFunc: func(pod *api.Pod) (*api.PodStatus, error) {
```
this is more involved than the prior PodStatusFunc impl, why?

jdef replied:
nevermind, I see why

jdef replied:
this still feels dirty, doing locking here because of the order of init

sttts replied:
It is dirty. Explicitly dirty ;-) I want to get rid of this whole callback mechanism. Why can't we just use the apiserver to get this information?

My mid-term goal is to run the executor and the kubelet without any need to synchronize their start-up sequence.

jdef replied:
this func is what provides information to the apiserver.

On Tue, Sep 8, 2015 at 10:15 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r38930044>
> :
>
> > +	execUpdates := make(chan kubelet.PodUpdate, 1)
> > +	exec := executor.New(executor.Config{
> > +		Updates:         execUpdates,
> > +		APIClient:       apiclient,
> > +		Docker:          dockerClient,
> > +		SuicideTimeout:  s.SuicideTimeout,
> > +		KubeletFinished: kubeletFinished,
> > +		ShutdownAlert: func() {
> > +			if shutdownCloser != nil {
> > +				if e := shutdownCloser.Close(); e != nil {
> > +					log.Warningf("failed to signal shutdown to external watcher: %v", e)
> > +				}
> > +			}
> > +		},
> > +		ExitFunc: os.Exit,
> > +		PodStatusFunc: func(pod *api.Pod) (*api.PodStatus, error) {
>
> It is dirty. Explicitly dirty ;-) I want to get rid of this whole callback
> mechanism. Why can't we just use the apiserver to get this information?
>
> My mid-term goal is to run the executor and the kubelet without any need
> to synchronize their start-up sequence.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r38930044>.
>

sttts replied:
Isn't the status manager not running inside the kubelet?

sttts replied:
Actually it is according to the logs: 
```
status_manager.go:77] Starting to sync pod status with apiserver
```

jdef commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -75,8 +75,8 @@ const (
-	defaultExecutorCPUs      = mresource.CPUShares(0.25)  // initial CPU allocated for executor
-	defaultExecutorMem       = mresource.MegaBytes(128.0) // initial memory allocated for executor
+	DefaultExecutorCPUs      = mresource.CPUShares(0.01)  // initial CPU allocated for executor, intentionally set to the minimum allow by Mesos: 0.01
+	DefaultExecutorMem       = mresource.MegaBytes(128.0) // initial memory allocated for executor
```
these defaults shouldn't be needed outside of this package

jdef commented:
There's a bit of a chicken-and-egg problem if we're relying upon the executor/kubelet to apply slave attribute labels to the node objects: the executor/kubelet don't get launched until there are tasks. and if tasks depend on nodes w/ specific labels in order to get launched then an executor/kubelet may never manifest on a node in the first place.

sttts commented:
Don't think the chicken-egg-problem exists (luckily). Our scheduler does not schedule to nodes, but to Mesos slaves with attributes. Hence, the scheduler knows the labels already before an executor+kubelet combo is started.

jdef commented:
Good point. I was thinking that the scheduler used node labels directly.

On Thu, Sep 10, 2015 at 12:06 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> Don't think the chicken-egg-problem exists (luckily). Our scheduler does
> not schedule to nodes, but to Mesos slaves with attributes. Hence, the
> scheduler knows the labels already before an executor+kubelet combo is
> started.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036#issuecomment-139293980>
> .
>

sttts commented:
@jdef ptal. Removed the ugly executor.Done() construction.

jdef commented:
did this whole PR need review or were you looking for eyes on the last
commit only?

On Tue, Sep 22, 2015 at 2:56 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> @jdef <https://github.com/jdef> ptal. Removed the ugly executor.Done()
> construction.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036#issuecomment-142384466>
> .
>

sttts commented:
Since your last review only the last commit was added.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -179,6 +166,10 @@ func New(config Config) *KubernetesExecutor {
 	return k
 }
 
+func (k *KubernetesExecutor) InitialRegComplete() <-chan struct{} {
```
nit: can we just call this something simpler like `Registration()` and add a doc that says how to use the chan? e.g. `closes upon completion of the executor's Registered callback`

sttts replied:
I prefer RegistrationComplete. Registration is not intuitive. The channel is not a registration.

sttts replied:
or just Registered()

jdef replied:
unfortunately, Registered() is already taken :(

On Tue, Sep 22, 2015 at 4:49 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40141777>
> :
>
> > @@ -179,6 +166,10 @@ func New(config Config) *KubernetesExecutor {
> >  	return k
> >  }
> >
> > +func (k *KubernetesExecutor) InitialRegComplete() <-chan struct{} {
>
> or just Registered()
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40141777>.
>

sttts replied:
Registered is part of the interface. InitiallyRegistered makes sense. But compare with the comment below. We can maybe get rid of this.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -212,10 +199,15 @@ func (k *KubernetesExecutor) Registered(driver bindings.ExecutorDriver,
 	}
 
-	k.initialRegistration.Do(k.onInitialRegistration)
+	k.updateChan <- kubelet.PodUpdate{
```
possible (highly unlikely) data race between this chan write and the close of the chan in doShutdown (e.g. someone configures a very, very small suicide timeout on a heavily loaded system). solution would be to lock&defer-unlock the state mutex before writing to the chan here.

sttts replied:
We have more places where we write to the channel. Will add locking.

jdef replied:
The other places that we write to the channel should already be guarded by
locking (or at least document that they make that assumption)

2015-09-22 17:05 GMT-04:00 Dr. Stefan Schimanski <notifications@github.com>:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40144046>
> :
>
> >  	}
> >
> > -	k.initialRegistration.Do(k.onInitialRegistration)
> > +	k.updateChan <- kubelet.PodUpdate{
>
> We have more places where we write to the channel. Will add locking.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40144046>.
>

mikedanese replied:
no context at all on this but all operations (read, write, close) on channels are already threadsafe.

sttts replied:
With the little exception of writing to a closed channel which panics.

jdef replied:
writing and closing at the same time to a chan will cause a data race

2015-09-22 17:30 GMT-04:00 Dr. Stefan Schimanski <notifications@github.com>:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40146867>
> :
>
> >  	}
> >
> > -	k.initialRegistration.Do(k.onInitialRegistration)
> > +	k.updateChan <- kubelet.PodUpdate{
>
> With the little exception of writing to a closed channel which panics.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40146867>.
>

sttts replied:
I have added the missing locking. The other channel writes are fine already. But one has to do quite some control flow analysis to check that. This should be much more obvious.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -181,7 +190,75 @@ func (s *KubeletExecutorServer) Run(hks hyperkube.Interface, _ []string) error {
+		log.Info("executor Run completed")
+	}()
+
+	<-exec.InitialRegComplete()
```
maybe a comment here about how we're waiting here because we need to wait for pod sources to become established?

jdef replied:
also, is it possible for the executor to terminate before completion of registration? if so, we probably want a `select` here

sttts replied:
I wonder whether we can start to decouple even more by removing this wait. The initial reason was that we waited for the node labels from the registration SlaveInfo.

jdef commented:
minor comments, looking pretty good!

sttts committed:
Re-use KubeletServer.KubeletConfig and KubeletServer.Run

sttts committed:
Simplify executorKubelet decorator and move to kubelet.go

- simplify
- rename kubeletExecutor -> executorKubelet
- move executorKubelet code into kubelet.go
- remove duplicated executorKubelet.ListenAndServe
- fix executorKubelet.Run comment

sttts committed:
Decouple startup of executor and kubelet

sttts committed:
Re-use CreateAndInitKubelet

sttts commented:
@s-urbaniak ptal

s-urbaniak commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -93,43 +92,31 @@ type kuberTask struct {
 type KubernetesExecutor struct {
-	kl                   KubeletInterface   // the kubelet instance.
-	updateChan           chan<- interface{} // to send pod config updates to the kubelet
+	lock                 sync.RWMutex
```
This is a RWMutex, I don't see where we use `R[Un]Lock` in the code base below.

s-urbaniak commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -818,7 +771,10 @@ func (k *KubernetesExecutor) doShutdown(driver bindings.ExecutorDriver) {
+	k.lock.Lock()
+	close(k.terminate)
+	close(k.updateChan)
+	k.lock.Unlock()
```
Why do we need to lock in order to orchestrate closing those channels?

s-urbaniak commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -239,10 +219,16 @@ func (k *KubernetesExecutor) Registered(driver bindings.ExecutorDriver,
 
-	k.initialRegistration.Do(k.onInitialRegistration)
+	k.lock.Lock()
+	defer k.lock.Unlock()
```
what are we guarding here exactly?

sttts replied:
fixed.

sttts commented:
@s-urbaniak ptal. I added a RWMutex for the assumption that updateChan is still open. If you know a more elegant way for that, suggestions are welcome.

jdef commented:
I don't follow the lock refactoring in the executor at all -- why was this needed? My comment re: a potential race between chan writes and chan close should have required two line addition someplace. Instead there's a new lock and I'm somewhat confused by this approach.

sttts commented:
There was a version before which was wrong.

Basically for the time after `if isDone()` until a write to a channel we have a race because the channel could be closed meanwhile.

sttts commented:
If you see a more elegant solution for that, suggestions are welcome. I know it's ugly.

sttts commented:
To answer your comment "should have required two line addition someplace": you mean a lock around the code closing the channel. Then we would have to lock before the `isDone` call. This would be possible, but is essentially a mutex around all callbacks from the driver. Not nice either.

sttts commented:
And write-locking a read-locked mutex panics. So upgrading a read lock is not possible without a race.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -85,179 +71,162 @@ func NewKubeletExecutorServer() *KubeletExecutorServer {
+		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
+
+		// run the kubelet, until execUpdates is closed
+		err = s.KubeletServer.Run(kcfg)
```
we don't want a cloud provider to be initialized. invoking the upstream Run() doesn't give us that option.

sttts replied:
The cloud provider is only setup if nil is passed to Run.

jdef replied:
I don't understand this comment. I see nothing in the upstream Run() that
checks for nil relating to cloudprovider. What am I missing?

On Wed, Sep 23, 2015 at 10:19 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40208334>
> :
>
> >
> > -	hostNetworkSources, err := kubelet.GetValidatedSources(strings.Split(s.HostNetworkSources, ","))
> > +			for u := range execUpdates {
> > +				u.Source = MESOS_CFG_SOURCE
> > +				updates <- u
> > +			}
> > +		}()
> > +
> > +		// create static-pods directory file source
> > +		log.V(2).Infof("initializing static pods source factory, configured at path %q", staticPodsConfigPath)
> > +		fileSourceUpdates := kcfg.PodConfig.Channel(kubelet.FileSource)
> > +		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
> > +
> > +		// run the kubelet, until execUpdates is closed
> > +		err = s.KubeletServer.Run(kcfg)
>
> The cloud provider is only setup if nil is passed to Run.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40208334>.
>

sttts replied:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L424

jdef replied:
ugh. was looking at older/cached code - thanks

On Wed, Sep 23, 2015 at 10:44 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40211400>
> :
>
> >
> > -	hostNetworkSources, err := kubelet.GetValidatedSources(strings.Split(s.HostNetworkSources, ","))
> > +			for u := range execUpdates {
> > +				u.Source = MESOS_CFG_SOURCE
> > +				updates <- u
> > +			}
> > +		}()
> > +
> > +		// create static-pods directory file source
> > +		log.V(2).Infof("initializing static pods source factory, configured at path %q", staticPodsConfigPath)
> > +		fileSourceUpdates := kcfg.PodConfig.Channel(kubelet.FileSource)
> > +		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
> > +
> > +		// run the kubelet, until execUpdates is closed
> > +		err = s.KubeletServer.Run(kcfg)
>
>
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L424
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40211400>.
>

sttts commented:
@jdef Maybe we can just lock once during shutdown, without unlock ever again. Not sure that's nicer.

sttts committed:
Avoid panics during executor shutdown due to write to closed channel

s-urbaniak commented:
@sttts newest commits look much better now :+1: at least there is only one lock left :)

sttts commented:
Slightly squashed. Not finished yet.

sttts committed:
Pass PodLW to executor in tests

sttts commented:
Rebased, sorted and squashed.

sttts commented:
@jdef tests are green now. Feel free to take another look or give lgtm. It's only a rebase.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -165,12 +150,14 @@ func New(config Config) *KubernetesExecutor {
 
 	// watch pods from the given pod ListWatch
+	if config.PodLW == nil {
+		log.Fatal("cannot create executor with nil PodLW")
```
`panic()` here instead of `Fatal()` so that we can at least unit test this at some point

jdef replied:
.. and add docs that this func could potentially panic if config is missing things

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -257,38 +243,26 @@ func (k *KubernetesExecutor) Reregistered(driver bindings.ExecutorDriver, slaveI
-		Pods:   []*api.Pod{},
-		Op:     kubelet.SET,
-		Source: k.sourcename,
+		Pods: []*api.Pod{},
```
we probably shouldn't be sending this during re-registration. the first time an executor boots, yes. on slave fail-over? no.

sttts replied:
Rebase error. Good spot.

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -565,6 +542,9 @@ func (k *KubernetesExecutor) launchTask(driver bindings.ExecutorDriver, taskId s
 	k.pods[podFullName] = pod
 
 	// send the new pod to the kubelet which will spin it up
+	if k.isDone() {
```
instead of scattering calls to `k.isDone()` throughout the code, could we write something like this:
```
// sendPodUpdate assumes that caller is holding state lock; returns true when update is sent otherwise false
func sendPodUpdate(u *kubelet.PodUpdate) bool {
  if k.isDone() {
    return false
  }
  k.updateChan <- *u
  return true
}
```

sttts replied:
done

jdef commented on contrib/mesos/pkg/executor/service/kubelet.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+// runs the main kubelet loop, closing the kubeletFinished chan when the loop exits.
+// Returns when executorDone is closed.
```
this doesn't really return when executorDone is closed.

sttts replied:
That's the whole sense of our kubelet wrapper: the util.Until stops when executorDone is closed.

jdef replied:
the comment literally reads "Returns when executorDone is closed". the defer func() never returns. So Run(), in fact, never returns, right?

sttts replied:
Oh, now I got it. Of course, you are right. The select blocks. The whole construct only gives us the chance to close kubeletDone before blocking.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -62,6 +50,8 @@ type KubeletExecutorServer struct {
 	*app.KubeletServer
 	SuicideTimeout    time.Duration
 	LaunchGracePeriod time.Duration
+	kletLock          sync.Mutex
```
can we add a TODO to get rid of these ASAP?

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +76,164 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
+		kcfg.PodConfig = kconfig.NewPodConfig(kconfig.PodConfigNotificationIncremental, kcfg.Recorder) // override the default pod source
+		kcfg.StandaloneMode = false
+		kcfg.SystemContainer = "" // don't take control over other system processes.
+
```
I'd like to see us force-set kcfg's cloud-provider to nil

sttts replied:
I add a panic when it's set to fail hard and early.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +76,164 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
+	}
 	if err != nil {
-		return err
+		close(kubeletDone)
```
TODO: our kubelet.Run() override also closes this chan, be careful when refactoring

sttts replied:
added comments on both sides.

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +76,164 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
-	tlsOptions, err := s.InitializeTLS()
+	// create static pods directory
+	staticPodsConfigPath := filepath.Join(s.RootDirectory, "static-pods")
+	err := os.Mkdir(staticPodsConfigPath, 0755)
```
0755 is pretty open. what about 0750?

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +76,164 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
-		Writer:                         writer,
+	if err != nil {
+		// required for k8sm since we need to send api.Binding information back to the apiserver
+		log.Fatalf("No API client: %v", err)
```
let's return an error here instead

jdef commented:
left a few minor comments. super excited to see this almost ready to merge.

sttts committed:
Set static pods dir to 0750 instead of world readible 0755

sttts commented:
Squashed review fixes.

sttts committed:
Replace all executor_test timeouts by util.ForeverTestTimeout

sttts commented:
@jdef added a simple patch to fix some shippable flakes in executor_test. PTAL

jdef commented on contrib/mesos/pkg/executor/executor.go:
```diff
@@ -565,11 +540,10 @@ func (k *KubernetesExecutor) launchTask(driver bindings.ExecutorDriver, taskId s
 
 	// send the new pod to the kubelet which will spin it up
-	update := kubelet.PodUpdate{
+	k.sendPodUpdate(&kubelet.PodUpdate{
```
didn't the old code return here if we were terminating? how about:
```go
  ok := k.sendPodUpdate(...)
  if !ok {
    return
  }
```

jdef replied:
if we add this, should probably `delete(k.pods,podFullName)` prior to `return`

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +77,171 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
-		// required for k8sm since we need to send api.Binding information
-		// back to the apiserver
-		log.Fatalf("No API client: %v", err)
+		log.Fatalf("failed to create executor driver: %v", err)
```
`return fmt.Errorf(...)`

jdef commented on contrib/mesos/pkg/executor/service/service.go:
```diff
@@ -86,188 +77,171 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
+		if kcfg.Cloud != nil {
+			// fail early and hard because having the cloud provider loaded would go unnoticed,
+			// but break bigger cluster because accessing the state.json from every slave kills the master.
+			panic("cloud provider must not be set")
```
:+1:

jdef commented:
left a few more minor comments

sttts commented:
Rebased to catch up with upstream.

sttts commented:
Rebased to master with unbroken e2e tests.

sttts commented:
Squashed.

sttts commented:
Rebased because of "kubeletTypes -> kubetypes" renaming.

sttts commented:
EOF linefeed added in service.go.

sttts commented:
Go mergebot go

sttts commented:
Rebased to lastets master, hopefully with more stable e2e tests.

sttts commented:
@k8s-bot test this please

mikedanese commented:
The PR builder doesn't test these changes since they are only in contrib
